### DISCUSSION TITLE
Add LayerSkip to OWUI

### DISCUSSION THREAD

**Discussion by peuportier** [open-webui/open-webui#6342] [Ideas]
I do believe to integrate LayerSkip to OWUI will make a good impact. LayerSkip Speeds up LLM . We present LayerSkip, an end-to-end solution to speed-up inference of large language models (LLMs). First, during training we apply layer dropout, with low dropout rates for earlier layers and higher dropout rates for later layers, and an early exit loss where all transformer layers share the same exit. Second, during inference, we show that this training recipe increases the accuracy of early exit at earlier layers, without adding any auxiliary layers or modules to the model. Third, we present a novel self-speculative decoding solution where we exit at early layers and verify and correct with remaining layers of the model. Our proposed self-speculative decoding approach has less memory footprint than other speculative decoding approaches and benefits from shared compute and activations of the draft and verification stages. We run experiments on different Llama model sizes on different types of training: pretraining from scratch, continual pretraining, finetuning on specific data domain, and finetuning on specific task. We implement our inference solution and show speedups of up to 2.16x on summarization for CNN/DM documents, 1.82x on coding, and 2.0x on TOPv2 semantic parsing task. 

https://github.com/facebookresearch/LayerSkip

**Comment by mostafaelhoushi** [comment]
Hi. Author of LayerSkip here. Happy to provide guidance to integrate it into OWUI.
If OWUI has support for speculative decoding, then integrating LayerSkip should be quick and easy.

**Comment by peuportier** [comment]
@mostafaelhoushi Hello my friend. In the first place, thanks for you amazing job. LayerSkip is insane good sh.. bro. Super amazing. For OWUI it would be cool to be integrate with OLLAMA as model LLM provider. That would make a significant advancement. @tjbck I know that is not yet on your agenda. Due to amount of work you have. But may it be consider, especially if the Author here offer contribution? Thanks you a lot anyway for all the contribution and this fantastic tool that is OWUI.

**Reply by mostafaelhoushi** [reply]
Thanks @peuportier for your kind words. 
Regarding the quickest way to get LayerSkip working, is to leverage an existing speculative decoding implementation (if it exists on OLLAMA) by a doing a trick similar to what we did with HuggingFace:
```
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch
from copy import deepcopy

checkpoint = "facebook/layerskip-llama2-7B"
early_exit = 4
device = "cuda" if torch.cuda.is_available() else "cpu"
prompt = "typing import List\ndef bucket_sort(A: List):"

model = AutoModelForCausalLM.from_pretrained(checkpoint)
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

generation_config = model.generation_config

# This is the trick here to create a draft model that shares the weights with the main model
weights_memo = {id(w): w for w in model.parameters()}
assistant_model = deepcopy(model, memo=weights_memo) # Clone main model with shared weights
assistant_model.model.layers = assistant_model.model.layers[:early_exit] # Apply early exit
del assistant_model.model.layers[early_exit:]

model.to(device)
assistant_model.to(device)

inputs = tokenizer(prompt, return_tensors="pt").to(device)

# here we leverage HuggingFace's existing speculative decoding support by passing in the early exit version of the model into `assistant_model` arg
outputs = model.generate(**inputs, assistant_model=assistant_model, generation_config=generation_config)
print(tokenizer.batch_decode(outputs, skip_special_tokens=True)[0])
```

This quick trick will get things working and it shares weights of the draft and main stage.

There are other optimizations to share KV cache and share the compute of the earlier layers that are done [here](https://github.com/pytorch-labs/gpt-fast/compare/6253c6bb054e658d67566150f87329b87815ae63...417aca13cf16f68b0f76fa1b9fd44da7470b7c59) on GPT-Fast but we can discuss that later.

**Reply by peuportier** [reply]
@mostafaelhoushi My friend, Maybe we shall ask Ollama community about this. To integrate your solution into Ollama then after OWUI will benefit from it. if can accelerate model. That would be best key
